lib/os/onoff.c

```diff
@@ -1,127 +1,81 @@
 
-static void set_service_state(struct onoff_service *srv,
+static void set_service_state(struct onoff_manager *mgr,
 			      u32_t state)
 {
-	srv->flags &= ~SERVICE_STATE_MASK;
-	srv->flags |= (state & SERVICE_STATE_MASK);
+	mgr->flags &= ~ST_MASK;
+	mgr->flags |= (state & ST_MASK);
 }
 
-static int validate_args(const struct onoff_service *srv,
+static int validate_args(const struct onoff_manager *mgr,
 			 struct onoff_client *cli)
 {
-	if ((srv == NULL) || (cli == NULL)) {
+	if ((mgr == NULL) || (cli == NULL)) {
 		return -EINVAL;
 	}
 
-	int rv = 0;
-	u32_t mode = cli->flags;
+	int rv = async_notify_validate(&cli->notify);
 
-	/* Reject unexpected flags. */
-	if (mode != (cli->flags & CLIENT_VALID_FLAGS_MASK)) {
-		return -EINVAL;
-	}
-
-	/* Validate configuration based on mode */
-	switch (mode & CLIENT_NOTIFY_METHOD_MASK) {
-	case ONOFF_CLIENT_NOTIFY_SPINWAIT:
-		break;
-	case ONOFF_CLIENT_NOTIFY_CALLBACK:
-		if (cli->async.callback.handler == NULL) {
-			rv = -EINVAL;
-		}
-		break;
-	case ONOFF_CLIENT_NOTIFY_SIGNAL:
-		if (cli->async.signal == NULL) {
-			rv = -EINVAL;
-		}
-		break;
-	default:
+	if ((rv == 0)
+	    && ((cli->notify.flags & ASYNC_NOTIFY_EXTENSION_MASK) != 0)) {
 		rv = -EINVAL;
-		break;
-	}
-
-	/* Clear the result here instead of in all callers. */
-	if (rv == 0) {
-		cli->result = 0;
 	}
 
 	return rv;
 }
 
-int onoff_service_init(struct onoff_service *srv,
-		       onoff_service_transition_fn start,
-		       onoff_service_transition_fn stop,
-		       onoff_service_transition_fn reset,
-		       u32_t flags)
+int onoff_manager_init(struct onoff_manager *mgr,
+		       const struct onoff_transitions *transitions)
 {
-	if ((flags & SERVICE_CONFIG_FLAGS) != flags) {
+	if (transitions->flags & ~SERVICE_CONFIG_FLAGS) {
 		return -EINVAL;
 	}
 
-	if ((start == NULL) || (stop == NULL)) {
+	if ((transitions->start == NULL) || (transitions->stop == NULL)) {
 		return -EINVAL;
 	}
 
-	*srv = (struct onoff_service)ONOFF_SERVICE_INITIALIZER(start, stop,
-							       reset, flags);
+	*mgr = (struct onoff_manager)ONOFF_MANAGER_INITIALIZER(transitions);
 
 	return 0;
 }
 
-static void notify_one(struct onoff_service *srv,
+static void notify_one(struct onoff_manager *mgr,
 		       struct onoff_client *cli,
 		       int res)
 {
-	unsigned int flags = cli->flags;
+	onoff_client_callback cb =
+		(onoff_client_callback)async_notify_finalize(&cli->notify, res);
 
-	/* Store the result, and notify if requested. */
-	cli->result = res;
-	cli->flags = 0;
-	switch (flags & CLIENT_NOTIFY_METHOD_MASK) {
-	case ONOFF_CLIENT_NOTIFY_SPINWAIT:
-		break;
-	case ONOFF_CLIENT_NOTIFY_CALLBACK:
-		cli->async.callback.handler(srv, cli,
-					    cli->async.callback.user_data, res);
-		break;
-#ifdef CONFIG_POLL
-	case ONOFF_CLIENT_NOTIFY_SIGNAL:
-		k_poll_signal_raise(cli->async.signal, res);
-		break;
-#endif /* CONFIG_POLL */
-	default:
-		__ASSERT_NO_MSG(false);
+	if (cb) {
+		cb(mgr, cli, cli->user_data, res);
```

Comment:
there was a concern that `user_data` might be overwritten by this time because of fetch result reports that client is completed after returning from `async_notify_finalize`. You mentioned that it will be documented https://github.com/zephyrproject-rtos/zephyr/pull/23229#discussion_r391646030. Can you point me to that? Anyway, i think that it would be safer to store `user_data` before calling `async_notify_finalize`

Reply:
Done. For this API I can't see that it hurts, other than suggesting that doing this is sufficient to avoid the problem of repurposing the notification structure.  If somebody provides a callback that needs additional context from a containing object that won't have been cached.